### PR TITLE
Add Snapshot location to compute snapshot

### PR DESCRIPTION
**Pull request by upodroid**
Fixes: https://github.com/hashicorp/terraform-provider-google/issues/6941

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [ ] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: added storage_locations field to `google_compute_snapshot`
```

### CODE REVIEW

File: products/compute/api.yaml [hunk 1]
```diff
@@ -12745,6 +12755,12 @@ objects:
           creation/deletion.
         output: true
       # 'storageBytesStatus' not useful for object convergence.
+      - !ruby/object:Api::Type::Array
+        name: 'storageLocations'
+        description: |
+          Cloud Storage bucket storage location of the snapshot (regional or multi-regional).
+        item_type: Api::Type::String
+        output: true
```
Comment (megan07) [inline products/compute/api.yaml:12763]: Should this be `output:true`? It looks to be configurable.

Comment (upodroid) [reply products/compute/api.yaml:12763]: fixed

**Comment by upodroid** [comment]
I didn't write a test for the KMS fields. I don't currently see one using CMEK, there is a CSEK test though.

**Comment by megan07** [comment]
Hi @upodroid! Thanks again for this contribution - and I apologize for my delay in responding, I wanted to make sure I had ample time to review it.

As for tests, I think you can probably add tests to `resource_compute_snapshot_test.go` - for examples on how to use the kms key, take a look at `TestAccComputeDisk_encryptionKMS` maybe?

Also, I was playing around with it a little and I think if `storage_locations` is not set, there is a default returned by the API, so we might want to make that Optional + Computed.

Let me know if you have more questions! Thanks! And sorry again for the delay!

**Comment by upodroid** [comment]
@megan07 The tests are kind of passing now.

The APIs are being a bit dodgy. The service account was used to encrypt the disk first and then create the snapshot. The CMEK works for the disk creation but fails the snapshot creation even thought it happens after the disk is created.

I tried it manually via the APIs and it works correctly. Terraform is a bit flaky.

```
 REDACTED  MCW0CDP3YY  ~  go  …  github.com  hashicorp  terraform-provider-google-beta   master  29⬇  8✎  8+  USAGE  $    make testacc TEST=./google-beta TESTARGS='-run=TestAccComputeSnapshot_encryptionCMEK'
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google-beta -v -run=TestAccComputeSnapshot_encryptionCMEK -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google-beta/version.ProviderVersion=acc"
=== RUN   TestAccComputeSnapshot_encryptionCMEK
=== PAUSE TestAccComputeSnapshot_encryptionCMEK
=== CONT  TestAccComputeSnapshot_encryptionCMEK
    TestAccComputeSnapshot_encryptionCMEK: testing.go:674: Step 0 error: errors during apply:
        
        Error: Error creating Snapshot: googleapi: Error 400: Cloud KMS error when using key projects/REDACTED/locations/us-central1/keyRings/tf-test-tuc2a1pnxq/cryptoKeys/tf-test-tuc2a1pnxq/cryptoKeyVersions/1: Permission 'cloudkms.cryptoKeyVersions.useToDecrypt' denied on resource 'projects/REDACTED/locations/us-central1/keyRings/tf-test-tuc2a1pnxq/cryptoKeys/tf-test-tuc2a1pnxq' (or it may not exist)., kmsPermissionDenied
        
          on /var/folders/b1/dthn83bs2qbcrg38qszm22440000gn/T/tf-test534309074/main.tf line 39:
          (source code not available)
        
        
--- FAIL: TestAccComputeSnapshot_encryptionCMEK (56.02s)
FAIL
FAIL    github.com/hashicorp/terraform-provider-google-beta/google-beta 60.536s
FAIL
make: *** [testacc] Error 1
```

**Review by megan07 — CHANGES_REQUESTED** [review]
After making a few of the suggested changes, I ran the tests locally and ran into a couple other things - the `Disk` resource has a `decoder` (`templates/terraform/decoders/disk.erb`) that decodes the encryption keys, the child attributes are hard-coded in that decoder, so now that we've added `kmsKeyServiceAccount` to it's children, we'll need to make sure it's added there as well.

Along the same lines, you'll see in that decoder that one of the things decoded is that the `kmsKeyName` actually returns the version, so we handle that in there - and we need to do the same thing in a decoder for `Snapshot` now as well.

Let me know if you have any questions, but I think this should get us close! Thanks for all of your work put in here!

**Comment by megan07** [comment]
One other thing, as I was looking into the code locally, I saw some extra whitespace at the end of lines, would you mind cleaning that up a bit too, please?  Thanks!

File: third_party/terraform/tests/resource_compute_snapshot_test.go [hunk 2]
```diff
@@ -63,3 +87,54 @@ resource "google_compute_snapshot" "foobar" {
...
+  
+resource "google_kms_crypto_key" "example-key" {
+  name            = "%s"
+  key_ring        = google_kms_key_ring.keyring.id
+  rotation_period = "100000s"
+}
+
+resource "google_kms_crypto_key_iam_member" "example-key" {
+  crypto_key_id = google_kms_crypto_key.example-key.id
+  role          = "roles/cloudkms.cryptoKeyEncrypter"
```
[the remark this answers is not in the document]
Comment (upodroid) [reply third_party/terraform/tests/resource_compute_snapshot_test.go:114]: rookie mistake 😄

**Comment by upodroid** [comment]
This is working now. That decoder hint gave me ideas on how to fix PR 3939.

```
REDACTED  MCW0CDP3YY  ~  go  …  github.com  hashicorp  terraform-provider-google-beta   master  37⬇  10✎  USAGE  $    make testacc TEST=./google-beta TESTARGS='-run=TestAccComputeSnapshot_encryptionCMEK'
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google-beta -v -run=TestAccComputeSnapshot_encryptionCMEK -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google-beta/version.ProviderVersion=acc"
=== RUN   TestAccComputeSnapshot_encryptionCMEK
=== PAUSE TestAccComputeSnapshot_encryptionCMEK
=== CONT  TestAccComputeSnapshot_encryptionCMEK
--- PASS: TestAccComputeSnapshot_encryptionCMEK (91.82s)
PASS
ok      github.com/hashicorp/terraform-provider-google-beta/google-beta 97.315s
```

File: products/compute/api.yaml [hunk 3]
```diff
@@ -2799,6 +2799,11 @@ objects:
             name: 'kmsKeyName'
             description: |
               The name of the encryption key that is stored in Google Cloud KMS.
+          - !ruby/object:Api::Type::String
```
Comment (megan07) [inline products/compute/api.yaml:2802]: Sorry, I missed this earlier. This looks like a duplicate key though. Otherwise, I think it's looking good!

Comment (upodroid) [reply products/compute/api.yaml]: fixed

**Review by megan07 — APPROVED** [review]
Looks good! Thank you!

**Pull request merged**